### PR TITLE
Closes #988: Create feature-download (Non UI).

### LINKED ISSUE #988: Create feature-downloads (Non UI)
Breaking in small parts this issue #459

- [ ] Subscribe to downloads in the currently selected session (I assume we do not need to subscribe to it in all Sessions?)

- [ ] Consume Download objects and pass the data to Androids Download Manager

- [ ] Call back into the app to notify if a download is done

### PR DESCRIPTION
**Pull request by Amejia481**
#988

### CODE REVIEW

File: components/feature/download/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# [Android Components](../../../README.md) > Feature > Download
+
+Feature implementation for apps that want to use Android downloads manager.
```
Comment (pocmo) [inline components/feature/download/README.md]: Let's link to the Android docs: [Android's Download Manager](https://developer.android.com/reference/android/app/DownloadManager)

File: components/feature/download/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+## Usage
```
Comment (pocmo) [inline components/feature/download/README.md]: Let's try to add some basic docs with new components: Let's describe what this actually does (forward downloads requests of the currently selected session to Android's Download Manager) and how to use it.

File: components/feature/download/build.gradle [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    testImplementation Deps.testing_robolectric
+    testImplementation Deps.testing_mockito
+    testImplementation project(':support-test')
+    testImplementation project(':concept-engine')
+}
+
+archivesBaseName = "feature-download"
+
+apply from: '../../../publish.gradle'
+apply plugin: 'kotlin-android-extensions'
```
Comment (pocmo) [inline components/feature/download/build.gradle]: nit: this should go to the top to the other plugins

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import android.content.Context.DOWNLOAD_SERVICE
+
+typealias OnDownloadFinish = (Long) -> Unit
+typealias AndroidDownloadManager = android.app.DownloadManager
+
+/**
+ * Handles the interactions with the [AndroidDownloadManager].
+ * @property context A reference to [Context] applicationContext.
+ */
+class DownloadManager(private val context: Context) {
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: nit: If we want this to be an applicationContext then let's make this a bit more obvious and name the parameter `applicationContext`.

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        val fileName = getFileName(download)
+
+        val request = Request(Uri.parse(download.url))
+            .setNotificationVisibility(VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+            .setMimeType(download.contentType)
+            .addRequestHeader("User-Agent", download.userAgent)
+
+        if (refererURL.isNotEmpty())
+            request.addRequestHeader("Cookie", cookie)
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: You are checking the `refererURL` and add the `cookie`. I guess we need to add both if they are not null?

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        val downloadID: Long
+
+        request.setDestinationInExternalPublicDir(download.destinationDirectory, fileName)
+
+        downloadID = androidDownloadManager.enqueue(request)
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: nit: It looks like downloadID can be created and assigned at the same place:

```Kotlin
val downloadID = androidDownloadManager.enqueue(request)
```

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        queuedDownloads[downloadID] = downloadListener
+
+        if (!isSubscribedReceiver)
+            registerBroadcastReceiver(context)
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: nit: Let's always wrap blocks in `{ }`. We may want to enable a detekt rule for that.

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    private fun getFileName(download: Download): String? {
+        return if (!download.fileName.isNullOrEmpty())
+            download.fileName
+        else
+            DownloadUtils.guessFileName(
+                "",
+                download.url,
+                download.contentType
+            )
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: I think we should file a follow-up to make sure fileName is never null. No reason to let consumers call `guessFileName`. And I think in `browser-engine-system` we already do that. We just need to make sure that the gecko-flavors do that too.

File: samples/firefox-accounts/build.gradle [hunk 9]
```diff
@@ -38,7 +38,7 @@ android {
 dependencies {
     implementation project(':service-firefox-accounts')
 
-    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
+    implementation Deps.support_constraintlayout
```
Comment (pocmo) [inline samples/firefox-accounts/build.gradle]: nit: unrelated change. Let's always file a separate issue for that to not make the review diff too large.

File: components/feature/download/README.md (hunk 1 above)
Comment (Amejia481) [reply components/feature/download/README.md]: Should I also update the diagram in the main README.md?
I noticed that Intent and Tabs features aren't there too.

image

Comment (Amejia481) [reply components/feature/download/README.md]: 👍

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt (hunk 5 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: Oops my mistake, thanks for catching this one!

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt (hunk 8 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: I'm a bit confused, can you help me to clarify please.

1.Do you want that I change Download.fileName property to Non-null? **e.g fileName: String** instead of  **fileName: String?** ?

2.Or are you referring to `private fun getFileName(download: Download): String?` type return  **String?**

Comment (pocmo) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: Option 1: We should always come up with a file name. But that's something for a separate issue.

Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: https://github.com/mozilla-mobile/android-components/issues/1017

**Comment by Amejia481** [comment]
I've added all the suggestion to the PR!

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+
+/**
+ * Feature implementation for proving download functionality for the selected session.
+ */
+class DownloadManagerFeature(
+    private val downloadManager: DownloadManager,
+    private val sessionManager: SessionManager
+) : SessionManager.Observer, Session.Observer {
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: As there is only one `DownloadManager` how about adding that as default parameter:

```Kotlin
class DownloadManagerFeature(
    private val sessionManager: SessionManager,
    private val downloadManager: DownloadManager = DownloadManager()
) : SessionManager.Observer, Session.Observer {
```

If you see the need for multiple download manager implementations in the future (I guess that makes sense) then I'd expect `DownloadManager` to be an interface and your implementation to have a different name (`AndroidDownloadManager` - but then your `typealias` may break 😜).

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /**
+     * Notifies to the [DownloadManager] that a new download must be processed.
+     */
+    @SuppressLint("MissingPermission")
+    override fun onDownload(session: Session, download: Download): Boolean {
+        downloadManager.download(download)
+        return true
+    }
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: We need a way to let the app request the permission only when needed.

I'm thinking of something like this:
* `onDowload()` is getting invoked
* If we do **not** have the permissions:
   * let the app know via a callback that it needs to prompt
   * Return false here so that the `Download` object will be retained
   * Add a method that the app can call when a permission was granted
   * Start the download if we have all permissions now
* If we do have the permissions:
   * Start the download immediately

If this is a too big change then let's work on that in a follow-up issue. We will need that to use such a component in an app like Focus. None of the apps will want to ask the user upfront to accept those permissions. At the time of a download coming in is the best in-context place to ask.

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * Remove all the listeners.
+     */
+    fun unregisterListeners() {
+        applicationContext.unregisterReceiver(onDownloadComplete)
+        isSubscribedReceiver = false
+        queuedDownloads.clear()
+    }
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: Your listener setup is very fine-grained and I wonder if we can do it simpler. The apps we have are not really interested in subscribing to specific downloads. For example Focus shows a `Snackbar` whenever a download is completed. For that some simple callback would be enough:

```
onDownloadCompleted(download: Download, ... additional meta data if needed)
```

This callback (if registered) could get invoked for all downloads and I think I'd want to register that on the `Feature` when initializing it.

Does this make sense?

File: components/feature/download/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        )
+
+        /*Adding a listener to download in process
+        * If the download I've already had a listener
+        * it will be updated with the new one*/
+        manager.registerListener(downloadId, onDownloadComplete)
+
+        /*Unregistering all listeners*/
+        manager.unregisterListeners()
+```
```
Comment (pocmo) [inline components/feature/download/README.md]: nit: I see the `DownloadManager` class more as an implementation detail. Do you think there's a need to use it in a standalone way? Anyhow, I'd at least move this to the end and explain `DownloadManagerFeature` first. :)

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import android.annotation.SuppressLint
+import android.support.annotation.RequiresPermission
+import mozilla.components.browser.session.Download
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+
+/**
+ * Feature implementation for proving download functionality for the selected session.
+ */
+class DownloadManagerFeature(
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: nit: I think for consistency it would be nice to use the same name as the module: `DownloadFeature`.

Originally I was thinking of `feature-downloads` (s) and `DownloadsFeature`. But the *s* is not that important :D

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+     * @param downloadID the id that reference to the schedule download.
+     * @param listener a listener to be notify when the download is completed.
+     */
+    @Suppress("unused")
+    fun registerListener(downloadID: Long, listener: OnDownloadFinish) {
+        if (downloadID in queuedDownloads)
+            queuedDownloads[downloadID] = listener
+        else
+            throw IllegalArgumentException("Unable to find $downloadID on the queued downloads")
+    }
```
Comment (pocmo) [inline components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: nit: There are multiple if statements here and above without `{ }`

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt [hunk 16]
```diff
@@ -33,3 +35,11 @@ fun Context.isOSOnLowMemory(): Boolean {
...
     }.lowMemory
 }
+
+/**
+ * Returns if a list of permission have been granted, if all the permission have been granted
+ * returns true otherwise false.
+ */
+fun Context.isPermissionGranted(vararg permission: String): Boolean {
+    return permission.all { checkSelfPermission(this, it) == PERMISSION_GRANTED }
+}
```
Comment (pocmo) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt]: Let's make sure to add this (and of course the new component!) to the changelog. :)

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt (hunk 10 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: Let's keep it simple for now :), I will the default parameter 😄

Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: But I will need an extra parameter (context) for creating a` DownloadManager(private val applicationContext: Context)`

Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: @pocmo it wont be worst to have an extra parameter, just for creating a default value. 

```kotlin
class DownloadManagerFeature(
    context:Context,
    private val downloadManager: DownloadManager = DownloadManager(context),
    private val sessionManager: SessionManager
) : SessionManager.Observer, Session.Observer
```

👍 
``` kotlin
       val feature = DownloadManagerFeature(context, sessionManager = sessionManager)
```

👎  the parameter is never been used
```kotlin
        val manager = DownloadManager(context)
        val feature = DownloadManagerFeature(context, manager, sessionManager)
```

Should I just add a context to the constructor and internally create the download manager? 

```kotlin
class DownloadManagerFeature(
    context: Context,
    private val sessionManager: SessionManager
) : SessionManager.Observer, Session.Observer {
    private val downloadManager: DownloadManager = DownloadManager(context)
```

Comment (pocmo) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: Ah, I see. Mh. Yeah, I think creating the `DownloadManager` internally may be a good workaround - as long as this doesn't affect testability? :)

Once we have the need to inject other DownloadManager things we can open that up again.

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt (hunk 11 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: What about this?
I add a new callback to the constructor `needToRequestPermissions`,
when I new download comes out and the app doesn't have the permissions I will call it.

When the app gets the permissions granted, the app can call the FeatureDownload.onDownload function with the same parameters that `needToRequestPermissions` gave it. 

What do you think?

```kotlin

class DownloadManagerFeature(
    private val context: Context,
    private val sessionManager: SessionManager,
    private val needToRequestPermissions: (session: Session, download: Download) -> Unit = { _, _ -> },
    private val downloadManager: DownloadManager = DownloadManager(context)
) : SessionManager.Observer, Session.Observer {

/*.....*/

 override fun onDownload(session: Session, download: Download): Boolean {
        return if (context.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE)) {
            downloadManager.download(download)
            true
        } else {
            needToRequestPermissions(session, download)
            false
        }
    }
}
```

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt (hunk 10 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: 👍

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt (hunk 11 above)
Comment (pocmo) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: I wonder if we can make it easier for the app and not require and parameters, something like: `onPermissionsGranted()`. Permission granting is asynchronously and it's quite some overhead for the app to keep a reference to the `Session` and `Download` across potential configuration changes etc.

Could we just check (in the callback) the selected session and see if it has a download attached and then re-trigger this ourselves?

Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: Got it! 👍

File: components/feature/download/README.md (hunk 13 above)
Comment (Amejia481) [reply components/feature/download/README.md]: Done!
I improved the documentation and put `DownloadsFeature` first.
I added the documentation of `DownloadManager` someone has a different use case that is not cover by `DownloadsFeature`.

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt (hunk 14 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt]: ^_^
I renamed the component to **feature-downloads** and `DownloadsFeature` 🕺 .

File: components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt (hunk 15 above)
Comment (Amejia481) [reply components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt]: Done `{ }` 👍

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt (hunk 16 above)
Comment (Amejia481) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt]: Done!
https://github.com/mozilla-mobile/android-components/pull/1032

File: components/support/test/build.gradle [hunk 17]
```diff
@@ -24,6 +24,7 @@ android {
 
     lintOptions {
         lintConfig file("lint.xml")
+        disable 'InvalidPackage'
```
Comment (Amejia481) [inline components/support/test/build.gradle]: I had to suppress this check to avoid  the lint error.
To be able to reuse Robolectric code.
image

Comment (pocmo) [reply components/support/test/build.gradle]: See lint.xml in the module. You can just list those jars to be excluded. There should already be one listed in the file.

Comment (Amejia481) [reply components/support/test/build.gradle]: 👍

Comment (Amejia481) [reply components/support/test/build.gradle]: Done! 🕺

**Comment by Amejia481** [comment]
@pocmo  should I integrate this one with  #1020?

**Comment by pocmo** [comment]
@Amejia481 Yeah, I think that would be a good idea. But it doesn't need to be now. Do you want to land that first or update the PR?

**Comment by Amejia481** [comment]
Ok, lets land this first 😄 !

**Comment by Amejia481** [comment]
@csadilek  helped me to add this feature and we did and less that 2 minutes. I'm updating the PR 😆

**Comment by pocmo** [comment]
Alright 🤣

File: components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt [hunk 18]
```diff
@@ -47,7 +47,7 @@ abstract class SelectionAwareSessionObserver(
     /**
      * Stops the observer.
      */
-    fun stop() {
+    open fun stop() {
```
Comment (Amejia481) [inline components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt]: We had to add this one to be able to override the method.

Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt]: Instead of extending `SelectionAwareSessionObserver` you could also create a local instance inside `DownloadsFeature`. May make it easier to understand what's going on too. :)

**Comment by pocmo** [comment]
>  sample browser

I meant reference browser! :)

File: samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt [hunk 19]
```diff
@@ -46,6 +51,13 @@ class BrowserFragment : Fragment(), BackHandler {
                 sessionId)
 
         tabsToolbarFeature = TabsToolbarFeature(context!!, toolbar, ::showTabs)
+
+        val ctx = context!!.applicationContext
```
Comment (pocmo) [inline samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt]: * nit: You can use `requireContext` in Fragments if you want to avoid `context!!` and are expecting to always have. a `Context`.
* nit: I feel the shorter `ctx` is harder to skim then just using `context` or `applicationContext` :)

Comment (Amejia481) [reply samples/browser/src/main/java/org/mozilla/samples/browser/BrowserFragment.kt]: Nice I didn't know about `requireContext()`, thanks for the suggestion!

File: components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt (hunk 18 above)
Comment (Amejia481) [reply components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt]: But it will also over-complicate things a bit.

Because then I will have to implement `Session.Observer` for getting notify on `onDownload` and I will need the `activeSession` and `SelectionAwareSessionObserver` just expose the `activeSession` for their children 😿  .

https://github.com/mozilla-mobile/android-components/blob/996c9c493beb21a182b08680889fa5df6e4b2a36/components/browser/session/src/main/java/mozilla/components/browser/session/SelectionAwareSessionObserver.kt#L23
